packages/opentelemetry-plugin-xml-http-request/src/xhr.ts

```diff
@@ -0,0 +1,371 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  B3Format,
+  BasePlugin,
+  isUrlIgnored,
+  isWrapped,
+  otperformance,
+} from '@opentelemetry/core';
+import * as tracing from '@opentelemetry/tracing';
+import * as types from '@opentelemetry/types';
+import {
+  addSpanNetworkEvent,
+  getResource,
+  parseUrl,
+  PerformanceTimingNames as PTN,
+  sameOriginOrUrlMatches,
+} from '@opentelemetry/web';
+import * as shimmer from 'shimmer';
+import { AttributeNames } from './enums/AttributeNames';
+import { EventNames } from './enums/EventNames';
+import {
+  OpenFunction,
+  PropagateTraceHeaderUrls,
+  SendBody,
+  SendFunction,
+  XMLHttpRequestWrapped,
+} from './types';
+
+/**
+ * XMLHttpRequest config
+ */
+export interface XMLHttpRequestPluginConfig extends types.PluginConfig {
+  // urls which should include trace headers when origin matches
+  propagateTraceHeaderUrls?: PropagateTraceHeaderUrls;
+}
+
+/**
+ * This class represents a XMLHttpRequest plugin for auto instrumentation
+ */
+export class XMLHttpRequestPlugin extends BasePlugin<XMLHttpRequest> {
+  readonly component: string = 'xml-http-request';
+  readonly version: string = '0.3.0';
+  moduleName = this.component;
+
+  protected _config!: XMLHttpRequestPluginConfig;
+
+  private _callbackToRemoveEvents: { [key: string]: Function } = {};
+  private _spans: { [key: string]: tracing.Span } = {};
+  private _usedResources: { [key: string]: PerformanceResourceTiming[] } = {};
+
+  constructor(config: XMLHttpRequestPluginConfig = {}) {
+    super();
+    this._config = config as XMLHttpRequestPluginConfig;
+  }
+
+  /**
+   * Adds custom headers to XMLHttpRequest
+   * @param xhr
+   * @param span
+   * @private
+   */
+  private _addHeaders(xhr: XMLHttpRequest, span: types.Span) {
+    let propagateTraceHeaderUrls = this._config.propagateTraceHeaderUrls;
+    if (!propagateTraceHeaderUrls) {
+      return;
+    }
+
+    if (
+      typeof propagateTraceHeaderUrls === 'string' ||
+      propagateTraceHeaderUrls instanceof RegExp
+    ) {
+      propagateTraceHeaderUrls = [propagateTraceHeaderUrls];
+    }
+
+    const url = (span as tracing.Span).name;
+    const spanUrl = parseUrl(url);
+    if (spanUrl.origin !== window.location.origin) {
+      this._logger.warn('Cannot set headers on different origin');
```

Comment:
So my thinking is that the `propagateTraceHeaderUrls` is what allows us to choose to send trace headers to different origins, such that if the user explicitly opts-in we will send headers to that origin.

On the other hand, if it's same origin, I think we should just always propagate the trace headers.